assets/js/util/tracking/createTrackEvent.js

```diff
@@ -32,17 +33,49 @@ export default function createTrackEvent( config, dataLayerTarget ) {
 		} = config;
 
 		if ( ! trackingEnabled ) {
-			return;
+			// Resolve immediately if tracking is disabled.
+			return Promise.resolve();
 		}
 
-		dataLayerPush( 'event', eventName, {
-			send_to: trackingID,
-			event_category: eventCategory,
-			event_label: eventLabel,
-			event_value: eventValue,
-			dimension1: referenceSiteURL,
-			dimension2: isFirstAdmin ? 'true' : 'false',
-			dimension3: userIDHash,
+		return new Promise( ( resolve ) => {
+			const eventData = {
+				send_to: trackingID,
+				event_category: eventCategory,
+				event_label: eventLabel,
+				event_value: eventValue,
+				dimension1: referenceSiteURL,
+				dimension2: isFirstAdmin ? 'true' : 'false',
+				dimension3: userIDHash,
+			};
+
+			let resolved = false;
+			const resolveOnce = ( eventSent ) => {
+				if ( resolved ) {
+					return;
+				}
+				resolved = true;
+				if ( ! eventSent ) {
+					global.console.warn( `Tracking event "${ eventName }" (category "${ eventCategory }") took too long to fire.` );
+				}
+				resolve();
+			};
```

Comment:
This is actually not to call `resolve` once as much as it is to only call the surrounding logic once - i.e. the warning shouldn't be triggered on the second call to `resolveOnce` (not part of the actual `resolve` function).

Reply:
In that case why not make a `warnOnce` function 😄 ?

You could simplify the implementation a bit using [`_.once`](https://lodash.com/docs/4.17.15#once)

```js
const warnOnce = once( global.console.warn )
```
Then `warnOnce` would only invoke the given function once per call to `trackEvent` since a new one would be created on each invocation.